gen/app-gen

```diff
@@ -0,0 +1,307 @@
+            fi
+            platform="$2"
+            shift 2
+            ;;
+        --image | -i)
+            # --image docker.io/library/debian:11,docker.io/library/debian:latest@sha256:a94cd7c7d58f483affd5937853ad4d24caa18cd7c2ec9ef65a9e528dfbc5eb07 --image docker.io/library/postgres:15.1
+            if [ -z "$2" ]; then
+                show_help_and_exit_error
+            fi
+            images+=($(echo "$2" | cut -f1 -d,)) # current
+            images+=($(echo "$2" | cut -f2 -d,)) # new, if current!=new then we need to generate delta.
+            shift 2
+            ;;
+        --device-type | -t)
+            if [ -z "$2" ]; then
+                show_help_and_exit_error
+            fi
+            device_types="$device_types $1 $2"
+            shift 2
+            ;;
+        --artifact-name | -n)
+            if [ -z "$2" ]; then
+                show_help_and_exit_error
+            fi
+            artifact_name=$2
+            shift 2
+            ;;
+        --output-path | -o)
+            if [ -z "$2" ]; then
+                show_help_and_exit_error
+            fi
+            output_path=$2
+            shift 2
+            ;;
+        -h | --help)
+            show_help
+            exit 0
+            ;;
+        --)
+            passthrough=1
+            shift
+            ;;
+        -*)
+            echo "Error: unsupported option $1"
+            show_help_and_exit_error
+            ;;
+        *)
+            shift
+            ;;
+    esac
+done
+
+if [ -z "${artifact_name}" ]; then
+    echo "Artifact name not specified. Aborting."
+    show_help_and_exit_error
+fi
+
+if [ -z "${device_types}" ]; then
+    echo "Device type not specified. Aborting."
+    show_help_and_exit_error
+fi
+
+if [ -z "${orchestrator}" ]; then
+    echo "Orchestrator not specified. Aborting."
+    show_help_and_exit_error
+fi
+
+if [ -z "${platform}" ]; then
+    echo "Platform not specified. Aborting."
+    show_help_and_exit_error
+fi
+
+if [ -z "${manifests_dir}" ]; then
+    echo "Directory containing manifests not specified. Aborting."
+    show_help_and_exit_error
+fi
+
+if [ ${#images[@]} -lt 1 ]; then
+    echo "No images specified. Aborting."
+    show_help_and_exit_error
+fi
+
+temp_dir=$(mktemp -d)
+[[ "${temp_dir}" == "" ]] && {
+    echo "cant get the temporary directory"
+    show_help_and_exit_error
+}
+mkdir "${temp_dir}/images"
+
+function cleanup() {
+    rm -Rf "$temp_dir"
+}
+trap cleanup EXIT SIGQUIT SIGTERM
+
+get_image_sha() {
+    local -r url="$1"
+
+    [[ "$url" == "" ]] && {
+        echo none
+        return 0
+    }
+    if [[ "${orchestrator}" == "docker-compose" ]]; then
+        docker image inspect "${url}" --format='{{.ID}}' | cut -f2 -d: # let's hope we do not have to preserve the fact that it was sha256:
+    else
+        ctr image ls name=="${url}" | sed -n -e 's/.*sha256:\([a-zA-Z0-9]*\).*/\1/p' # let's hope we do not have to preserve the fact that it was sha256:
```

Comment:
Nit: `ctr` not the most descriptive name I've seen thrown around :p

Reply:
in here [ctr](https://manpages.debian.org/bullseye/containerd/ctr.8.en.html#images,_image,_i) means the command line for containerd. This will be changed when add more orchestrators. probably only in case of `k8s` orchestrator it makes sense, although if you do not have docker, you can still use this and images will work. SInce atm we only have one orchestrator I left it like that.

Reply:
Ohh, TIL!

It says it's pretty much unstable though, are we sure we can rely on this going forwards?